edge-modules/api-proxy-module/rust-sdk/azure-iot-mqtt/src/io.rs

```diff
@@ -4,392 +4,518 @@ use std::future::Future;
+                match handshake.handshake() {
+                    Ok((mut stream, _)) => {
+                        stream.get_mut().set_cx(std::ptr::null_mut());
+                        std::task::Poll::Ready(Ok(stream))
+                    }
+
+                    Err(tungstenite::HandshakeError::Interrupted(mut handshake)) => {
+                        handshake.get_mut().get_mut().set_cx(std::ptr::null_mut());
+                        *self = WsConnect::Handshake(handshake);
+                        std::task::Poll::Pending
+                    }
+
+                    Err(tungstenite::HandshakeError::Failure(err)) => {
+                        poll_from_tungstenite_error(err)
+                    }
+                }
+            }
+
+            WsConnect::Invalid => panic!("future polled after completion"),
+        }
+    }
 }
 
 /// A wrapper around an inner I/O object
 pub enum Io<S> {
-	Raw(S),
+    Raw(S),
 
-	WebSocket {
-		inner: tungstenite::WebSocket<TokioToStd<S>>,
-		pending_read: std::io::Cursor<Vec<u8>>,
-	},
+    WebSocket {
+        inner: tungstenite::WebSocket<TokioToStd<S>>,
+        pending_read: std::io::Cursor<Vec<u8>>,
+    },
 }
 
-impl<S> tokio::io::AsyncRead for Io<S> where S: tokio::io::AsyncRead + tokio::io::AsyncWrite + Unpin {
-	fn poll_read(mut self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>, buf: &mut [u8]) -> std::task::Poll<std::io::Result<usize>> {
-		use std::io::Read;
-
-		let (inner, pending_read) = match &mut *self {
-			Io::Raw(stream) => return std::pin::Pin::new(stream).poll_read(cx, buf),
-			Io::WebSocket { inner, pending_read } => (inner, pending_read),
-		};
-
-		if buf.is_empty() {
-			return std::task::Poll::Ready(Ok(0));
-		}
-
-		loop {
-			if pending_read.position() != pending_read.get_ref().len() as u64 {
-				return std::task::Poll::Ready(Ok(pending_read.read(buf).expect("Cursor::read cannot fail")));
-			}
-
-			inner.get_mut().set_cx(cx);
-			let message = inner.read_message();
-			inner.get_mut().set_cx(std::ptr::null_mut());
-
-			match message {
-				Ok(tungstenite::Message::Binary(b)) => *pending_read = std::io::Cursor::new(b),
-				Ok(message) => log::warn!("ignoring unexpected message: {:?}", message),
-				Err(err) => return poll_from_tungstenite_error(err),
-			};
-		}
-	}
+impl<S> tokio::io::AsyncRead for Io<S>
+where
+    S: tokio::io::AsyncRead + tokio::io::AsyncWrite + Unpin,
+{
+    fn poll_read(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> std::task::Poll<std::io::Result<()>> {
+        use std::io::Read;
+
+        let (inner, pending_read) = match &mut *self {
+            Io::Raw(stream) => return std::pin::Pin::new(stream).poll_read(cx, buf),
+            Io::WebSocket {
+                inner,
+                pending_read,
+            } => (inner, pending_read),
+        };
+
+        if buf.filled().is_empty() {
```

Comment:
No, the intent of this check is to return early if the provided buffer has no remaining space to write to. So this needs to check `if buf.remaining() == 0`